parser/lexer.go

```diff
@@ -0,0 +1,499 @@
+				s.buf.WriteByte('\t')
+				s.r.inc()
+				continue
+			} else if ch0 == '\'' {
+				s.buf.WriteByte('\'')
+				s.r.inc()
+				continue
+			} else if ch0 == '_' || ch0 == '%' {
+				s.buf.WriteByte('\\')
+			}
+		} else if !isASCII(ch0) {
+			// TODO handle non-ascii
+		}
+		s.buf.WriteByte(ch0)
+		s.r.inc()
+	}
+
+	lit = s.buf.String()
+	// Quoted strings placed next to each other are concatenated to a single string.
+	// See http://dev.mysql.com/doc/refman/5.5/en/string-literals.html
+	ch := s.r.peek()
+	if ch == '\'' {
+		_, _, lit1 := s.scanString()
+		lit = lit + lit1
+	}
+
+	return
+}
+
+func (s *Scanner) scanNumber() (tok int, pos Pos, lit string) {
+	pos = s.r.pos()
+	ch0 := s.r.readByte()
+	switch ch0 {
+	case '0':
+		tok = intLit
+		ch1 := s.r.peek()
+		switch {
+		case ch1 >= '0' && ch1 <= '7':
+			s.r.inc()
+			s.scanOct()
+		case ch1 == 'x' || ch1 == 'X':
+			s.r.inc()
+			s.scanHex()
+			tok = hexLit
+		case ch1 == 'b':
+			s.r.inc()
+			s.scanBit()
+			tok = bitLit
+		case ch1 == '.':
+			return s.scanFloat(&pos)
+		case ch1 == 'B':
+			tok = unicode.ReplacementChar
+			return
+		}
+		lit = s.r.data(&pos)
+		return
+	case '.':
+		if isDigit(s.r.peek()) {
+			return s.scanFloat(&pos)
+		}
+		tok, lit = int('.'), "."
+		return
+	}
+
+	s.scanDigits()
+	ch0 = s.r.peek()
+	if ch0 == '.' || ch0 == 'e' || ch0 == 'E' {
+		return s.scanFloat(&pos)
+	}
+	tok, lit = intLit, s.r.data(&pos)
+	return
+}
+
+func (s *Scanner) scanOct() {
+	s.r.incAsLongAs(func(ch byte) bool {
+		return ch >= '0' && ch <= '7'
+	})
+}
+
+func (s *Scanner) scanHex() {
+	s.r.incAsLongAs(func(ch byte) bool {
+		return ch >= '0' && ch <= '9' ||
+			ch >= 'a' && ch <= 'f' ||
+			ch >= 'A' && ch <= 'F'
+	})
+}
+
+func (s *Scanner) scanBit() {
+	s.r.incAsLongAs(func(ch byte) bool {
+		return ch == '0' || ch == '1'
+	})
+}
+
+func (s *Scanner) scanFloat(beg *Pos) (tok int, pos Pos, lit string) {
+	s.r.p = *beg
+	// float = D1 . D2 e D3
+	s.scanDigits()
+	ch0 := s.r.peek()
+	if ch0 == '.' {
+		s.r.inc()
+		s.scanDigits()
+		ch0 = s.r.peek()
+	}
+	if ch0 == 'e' || ch0 == 'E' {
+		s.r.inc()
+		s.scanDigits()
+	}
+	tok, pos, lit = floatLit, *beg, s.r.data(beg)
+	return
+}
+
+func (s *Scanner) scanDigits() string {
+	pos := s.r.pos()
+	s.r.incAsLongAs(isDigit)
+	return s.r.data(&pos)
+}
+
+type reader struct {
+	s string
+	p Pos
+}
+
+var eof = Pos{-1, -1, -1}
+
+func (r *reader) eof() bool {
+	return r.p.Offset >= len(r.s)
+}
+
+func (r *reader) peek() byte {
+	if r.eof() {
+		return 0
+	}
+	return r.s[r.p.Offset]
+}
+
+func (r *reader) inc() {
+	if r.s[r.p.Offset] == '\n' {
+		r.p.Line++
+		r.p.Col = 0
+	}
+	r.p.Offset++
+	r.p.Col++
+}
+
+func (r *reader) incN(n int) {
+	for i := 0; i < n; i++ {
+		r.inc()
+	}
+}
+
+func (r *reader) readByte() (ch byte) {
+	ch = r.peek()
+	if ch == 0 && r.eof() {
+		return
+	}
+	r.inc()
+	return
+}
+
+func (r *reader) pos() Pos {
+	return r.p
+}
+
+func (r *reader) data(from *Pos) string {
+	return r.s[from.Offset:r.p.Offset]
```

Comment:
if that happens, it's a bug!
when you find any bugs, please report and I'll fix it.